File: client/old/src/main/java/net/treset/minecraftlauncher/launching/ResourceManager.java
package net.treset.minecraftlauncher.launching;

import net.treset.mc_version_loader.launcher.LauncherManifest;
import net.treset.minecraftlauncher.LauncherApplication;
import net.treset.minecraftlauncher.data.InstanceData;
import net.treset.minecraftlauncher.util.exception.GameResourceException;
import net.treset.minecraftlauncher.util.file.LauncherFile;
import net.treset.minecraftlauncher.util.string.PatternString;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.File;
import java.io.IOException;
import java.nio.file.StandardCopyOption;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ResourceManager {
    private static final Logger LOGGER = LogManager.getLogger(ResourceManager.class);
    
    private InstanceData instanceData;
    private boolean prepared = false;

    public ResourceManager(InstanceData instanceData) {
        this.instanceData = instanceData;
    }

    public void prepareResources() throws GameResourceException {
        addIncludedFiles(List.of(instanceData.getInstance().getKey(), instanceData.getOptionsComponent(), instanceData.getResourcepacksComponent(), instanceData.getSavesComponent()));
        if(instanceData.getModsComponent() != null) {
            addIncludedFiles(List.of(instanceData.getModsComponent().getKey()));
        }
        renameComponents();

        try {
            instanceData.setActive(true);
        } catch (IOException e) {
            throw new GameResourceException("Failed to prepare resources: unable to set instance active", e);
        }

        LOGGER.info("Prepared resources for launch, instance=" + instanceData.getInstance().getKey().getId());
    }

    public void setLastPlayedTime() throws IOException {
        LOGGER.debug("Setting last played time: instance={}", instanceData.getInstance().getKey().getId());
        if(instanceData.getInstance().getValue() == null) {
            throw new IOException("Unable to set last played time: instance details are null");
        }
        instanceData.getInstance().getValue().setLastPlayedTime(LocalDateTime.now());
        LauncherFile.of(instanceData.getInstance().getKey().getDirectory(), instanceData.getInstance().getKey().getDetails()).write(instanceData.getInstance().getValue());
        LOGGER.debug("Set last played time: instance={}", instanceData.getInstance().getKey().getId());
    }

    public void addPlayDuration(long duration) throws IOException {
        LOGGER.debug("Adding play duration: instance={}, duration={}", instanceData.getInstance().getKey().getId(), duration);
        if(instanceData.getInstance().getValue() == null) {
            throw new IOException("Unable to set last played time: instance details are null");
        }
        long oldTime = instanceData.getInstance().getValue().getTotalTime();
        instanceData.getInstance().getValue().setTotalTime(oldTime + duration);
        LauncherFile.of(instanceData.getInstance().getKey().getDirectory(), instanceData.getInstance().getKey().getDetails()).write(instanceData.getInstance().getValue());
        LOGGER.debug("Added play duration: instance={}, duration={}, totalTime={}", instanceData.getInstance().getKey().getId(), duration, instanceData.getInstance().getValue().getTotalTime());
    }

    public void cleanupGameFiles() throws GameResourceException {
        try {
            LOGGER.debug("Cleaning up game files: instance={}", instanceData.getInstance().getKey().getId());
            try {
                undoRenameComponents();
                ArrayList<LauncherFile> gameDataFilesList = getGameDataFiles();
                removeExcludedFiles(gameDataFilesList);
                removeIncludedFiles(List.of(instanceData.getSavesComponent()), gameDataFilesList);
                if (instanceData.getModsComponent() != null) {
                    removeIncludedFiles(List.of(instanceData.getModsComponent().getKey()), gameDataFilesList);
                }
                removeIncludedFiles(List.of(instanceData.getOptionsComponent(), instanceData.getResourcepacksComponent()), gameDataFilesList);
                removeRemainingFiles(gameDataFilesList);
            } catch (GameResourceException e) {
                throw new GameResourceException("Unable to cleanup game files", e);
            }

            try {
                instanceData.setActive(false);
            } catch (IOException e) {
                throw new GameResourceException("Unable to cleanup game files: unable to set instance inactive", e);
            }
        } catch (Exception e) {
            throw new GameResourceException("Unable to cleanup game files", e);
        }

        LOGGER.info("Game files cleaned up");
    }

    private void renameComponents() throws GameResourceException {
        LOGGER.debug("Renaming components: instance={}", instanceData.getInstance().getKey().getId());
        try {
            LauncherFile.of(instanceData.getSavesComponent().getDirectory()).moveTo(LauncherFile.of(instanceData.getGameDataDir(), "saves"), StandardCopyOption.ATOMIC_MOVE);
        } catch (IOException e) {
            throw new GameResourceException("Unable to rename saves file", e);
        }
        instanceData.getSavesComponent().setDirectory(LauncherFile.of(instanceData.getGameDataDir(), "saves").getPath());

        if(instanceData.getModsComponent() != null) {
            try {
                LauncherFile.of(instanceData.getModsComponent().getKey().getDirectory()).moveTo(LauncherFile.of(instanceData.getGameDataDir(), "mods"), StandardCopyOption.ATOMIC_MOVE);
            } catch (IOException e) {
                throw new GameResourceException("Unable to rename mods file", e);
            }
            instanceData.getModsComponent().getKey().setDirectory(LauncherFile.of(instanceData.getGameDataDir(), "mods").getPath());
        }
    }

    private void addIncludedFiles(List<LauncherManifest> components) throws GameResourceException {
        LOGGER.debug("Adding included files: instance={}", instanceData.getInstance().getKey().getId());
        List<GameResourceException> exceptionQueue = new ArrayList<>();
        for(LauncherManifest c : components) {
            if(c == null) {
                continue;
            }
            try {
                addIncludedFiles(c);
            } catch (GameResourceException e) {
                exceptionQueue.add(e);
                LOGGER.warn("Unable to get included files: manifestId=" + c.getId(), e);
            }
        }
        if(!exceptionQueue.isEmpty()) {
            throw new GameResourceException("Unable to get included files for " + exceptionQueue.size() + " components", exceptionQueue.get(0));
        }
        LOGGER.debug("Added included files: instance={}", instanceData.getInstance().getKey().getId());
    }

    private void addIncludedFiles(LauncherManifest manifest) throws GameResourceException {
        if(manifest == null || manifest.getIncludedFiles() == null) {
            throw new GameResourceException("Unable to get included files: unmet requirements");
        }
        LOGGER.debug("Adding included files: {}Id={}", manifest.getType(), manifest.getId());
        LauncherFile includedFilesDir = LauncherFile.of(manifest.getDirectory(), LauncherApplication.config.INCLUDED_FILES_DIR);
        if(!includedFilesDir.isDirectory()) {
            try {
                includedFilesDir.createDir();
            } catch (IOException e) {
                throw new GameResourceException("Unable to get included files: unable to create included files directory: manifestId=" + manifest.getId(), e);
            }
        }
        File[] files = includedFilesDir.listFiles();
        if(files == null) {
            throw new GameResourceException("Unable to get included files: manifestId=" + manifest.getId());
        }
        LOGGER.debug("Adding included files: {}Id={}, includedFiles={}", manifest.getType(), manifest.getId(), files);
        List<IOException> exceptionQueue = new ArrayList<>();
        for(File f : files) {
            LOGGER.debug("Adding included file: manifestId={}, file={}", manifest.getId(), f.getName());
            if(f.isFile() || f.isDirectory()) {
                try {
                    LauncherFile.of(f).copyTo(LauncherFile.of(instanceData.getGameDataDir(), f.getName()), StandardCopyOption.REPLACE_EXISTING);
                } catch (IOException e) {
                    exceptionQueue.add(e);
                    LOGGER.warn("Unable to move included files: unable to copy file: manifestId=" + manifest.getId());
                }
            } else {
                exceptionQueue.add(new IOException("Included files directory contains invalid file type: manifestId=" + manifest.getId()));
            }
        }
        if(!exceptionQueue.isEmpty()) {
            throw new GameResourceException("Unable to move included files: unable to copy " + exceptionQueue.size() + " files", exceptionQueue.get(0));
        }
        LOGGER.debug("Added included files: manifestId={}", manifest.getId());
    }

    private void undoRenameComponents() throws GameResourceException {
        LOGGER.debug("Undoing component renames: instance={}", instanceData.getInstance().getKey().getId());
        try {
            LauncherFile.of(instanceData.getSavesComponent().getDirectory()).moveTo(LauncherFile.of(instanceData.getGameDataDir(), instanceData.getSavesPrefix() + "_" + instanceData.getSavesComponent().getId()));
        } catch (IOException e) {
            throw new GameResourceException("Unable to cleanup launch resources: rename saves file failed", e);
        }
        instanceData.getSavesComponent().setDirectory(LauncherFile.of(instanceData.getGameDataDir(), instanceData.getSavesPrefix() + "_" + instanceData.getSavesComponent().getId()).getPath());

        if(instanceData.getModsComponent() != null) {
            try {
                LauncherFile.of(instanceData.getModsComponent().getKey().getDirectory()).moveTo(LauncherFile.of(instanceData.getGameDataDir(), instanceData.getModsPrefix() + "_" + instanceData.getModsComponent().getKey().getId()));
            } catch (IOException e) {
                throw new GameResourceException("Unable to cleanup launch resources: rename mods file failed", e);
            }
            instanceData.getModsComponent().getKey().setDirectory(LauncherFile.of(instanceData.getGameDataDir(), instanceData.getModsPrefix() + "_" + instanceData.getModsComponent().getKey().getId()).getPath());
        }
        LOGGER.debug("Undid component renames: instance={}", instanceData.getInstance().getKey().getId());
    }

    private ArrayList<LauncherFile> getGameDataFiles() throws GameResourceException {
        LOGGER.debug("Getting game data files: instance={}", instanceData.getInstance().getKey().getId());
        LauncherFile gameDataDir = instanceData.getGameDataDir();
        if(!gameDataDir.isDirectory()) {
            throw new GameResourceException("Unable to cleanup launch resources: game data directory not found");
        }

        File[] gameDataFiles = gameDataDir.listFiles();
        if(gameDataFiles == null) {
            throw new GameResourceException("Unable to cleanup launch resources: unable to get game data files");
        }

        LOGGER.debug("Got game data files: instance={}", instanceData.getInstance().getKey().getId());
        return new ArrayList<>(Arrays.stream(gameDataFiles).map(LauncherFile::of).toList());
    }

    private void removeExcludedFiles(ArrayList<LauncherFile> files) {
        LOGGER.debug("Removing excluded files: instance={}, files={}", instanceData.getInstance().getKey().getId(), files);
        List<LauncherFile> toRemove = new ArrayList<>();
        for(LauncherFile f : files) {
            if(f.getName().equals(LauncherApplication.config.MANIFEST_FILE_NAME) || PatternString.matchesAny(f.getName(), instanceData.getGameDataExcludedFiles())) {
                LOGGER.debug("Removing excluded file: instance={}, file={}", instanceData.getInstance().getKey().getId(), f.getName());
                toRemove.add(f);
            }
        }
        files.removeAll(toRemove);
        LOGGER.debug("Removed excluded files: instance={}", instanceData.getInstance().getKey().getId());
    }

    private void removeIncludedFiles(List<LauncherManifest> components, ArrayList<LauncherFile> files) throws GameResourceException {
        LOGGER.debug("Removing included files: instance={}, files={}", instanceData.getInstance().getKey().getId(), files);
        List<GameResourceException> exceptionQueue = new ArrayList<>();
        for(LauncherManifest component : components) {
            if(component.getIncludedFiles() == null || component.getIncludedFiles().isEmpty()) {
                LOGGER.debug("No included files: manifestId={}", component.getId());
                continue;
            }
            LOGGER.debug("Removing included files: manifestId={}", component.getId());
            try {
                removeIncludedFiles(component, files);
            } catch (GameResourceException e) {
                exceptionQueue.add(e);
                LOGGER.warn("Unable to remove included files: manifestId=" + component.getId(), e);
            }
        }
        if(!exceptionQueue.isEmpty()) {
            throw new GameResourceException("Unable to remove included files for " + exceptionQueue.size() + " components", exceptionQueue.get(0));
        }
        LOGGER.debug("Removed included files: instance={}", instanceData.getInstance().getKey().getId());
    }

    private void removeIncludedFiles(LauncherManifest component, ArrayList<LauncherFile> files) throws GameResourceException {
        LOGGER.debug("Removing included files: {}Id={}, includedFiles={}, files={}", component.getType(), component.getId(), component.getIncludedFiles(), files);
        LauncherFile includedFilesDir = LauncherFile.of(component.getDirectory(), LauncherApplication.config.INCLUDED_FILES_DIR);
        LauncherFile oldIncludedFilesDir = LauncherFile.of(component.getDirectory(), LauncherApplication.config.INCLUDED_FILES_DIR + "_old");
        if(includedFilesDir.exists()) {
            try {
                if(oldIncludedFilesDir.exists()) {
                    oldIncludedFilesDir.remove();
                }
                includedFilesDir.moveTo(oldIncludedFilesDir, StandardCopyOption.REPLACE_EXISTING);
            } catch (IOException e) {
                throw new GameResourceException("Unable to remove included files: unable to move included files directory: component_type=" + component.getType().name().toLowerCase() + " component=" + component.getId(), e);
            }
        }
        try {
            includedFilesDir.createDir();
        } catch (IOException e) {
            throw new GameResourceException("Unable to remove included files: unable to create included files directory: component_type=" + component.getType().name().toLowerCase() + " component=" + component.getId());
        }
        List<LauncherFile> toRemove = new ArrayList<>();
        List<IOException> exceptionQueue = new ArrayList<>();
        for(LauncherFile f : files) {
            String fName = f.isDirectory() ? f.getName() + "/": f.getName();
            boolean found = false;
            for(String i : component.getIncludedFiles()) {
                if(new PatternString(i).matches(fName)) {
                    LOGGER.debug("Removing included file: manifestId={}, file={}, matches={}", component.getId(), fName, i);
                    try {
                        f.moveTo(LauncherFile.of(component.getDirectory(), LauncherApplication.config.INCLUDED_FILES_DIR, f.getName()), StandardCopyOption.REPLACE_EXISTING);
                        found = true;
                    } catch (IOException e) {
                        exceptionQueue.add(e);
                        LOGGER.warn("Unable to remove included files: unable to move file: component_type=" + component.getType().name().toLowerCase() + " component=" + component.getId() + " file=" + f.getName());
                    }
                    break;
                }
            }
            if(found) {
                toRemove.add(f);
            } else {
                LOGGER.debug("Skipping included file: manifestId={}, file={}", component.getId(), f.getName());
            }
        }
        files.removeAll(toRemove);
        if(!exceptionQueue.isEmpty()) {
            throw new GameResourceException("Unable to remove included files: unable to move " + exceptionQueue.size() + " files: component_type=" + component.getType().name().toLowerCase() + " component=" + component.getId(), exceptionQueue.get(0));
        }
        LOGGER.debug("Removed included files: manifestId={}", component.getId());
    }

    private void removeRemainingFiles(ArrayList<LauncherFile> files) throws GameResourceException {
        LOGGER.debug("Removing remaining files: instance={}, files={}", instanceData.getInstance().getKey().getId(), files);
        LauncherFile includedFilesDir = LauncherFile.of(instanceData.getInstance().getKey().getDirectory(), LauncherApplication.config.INCLUDED_FILES_DIR);
        if(includedFilesDir.exists()) {
            try {
                includedFilesDir.remove();
            } catch (IOException e) {
                throw new GameResourceException("Unable to cleanup launch resources: unable to delete instance included files directory", e);
            }
        }
        if(!includedFilesDir.mkdir()) {
            throw new GameResourceException("Unable to cleanup launch resources: unable to create instance included files directory");
        }
        List<IOException> exceptionQueue = new ArrayList<>();
        for(LauncherFile f : files) {
            if(PatternString.matchesAny(f.getName(), instanceData.getInstance().getValue().getIgnoredFiles().stream().map(p -> new PatternString(p, true)).toList())) {
                LOGGER.debug("Deleting ignored file: instance={}, file={}", instanceData.getInstance().getKey().getId(), f.getName());
                try {
                    f.remove();
                } catch (IOException e) {
                    exceptionQueue.add(e);
                    LOGGER.warn("Unable to cleanup launch resources: unable to delete non-included file: " + f.getName());
                }
            }
            else {
                LOGGER.debug("Moving non-included file: instance={}, file={}", instanceData.getInstance().getKey().getId(), f.getName());
                try {
                    f.moveTo(LauncherFile.of(instanceData.getInstance().getKey().getDirectory(), LauncherApplication.config.INCLUDED_FILES_DIR, f.getName()), StandardCopyOption.REPLACE_EXISTING);
                } catch (IOException e) {
                    exceptionQueue.add(e);
                    LOGGER.warn("Unable to cleanup launch resources: unable to move non-included file: " + f.getName());
                }
            }
        }
        if(!exceptionQueue.isEmpty()) {
            throw new GameResourceException("Unable to cleanup launch resources: unable to cleanup " + exceptionQueue.size() + " non-included files", exceptionQueue.get(0));
        }
        LOGGER.debug("Removed remaining files: instance={}", instanceData.getInstance().getKey().getId());
    }

    public InstanceData getInstanceData() {
        return instanceData;
    }

    public void setInstanceData(InstanceData instanceData) {
        this.instanceData = instanceData;
    }

    public boolean isPrepared() {
        return prepared;
    }

    public void setPrepared(boolean prepared) {
        this.prepared = prepared;
    }
}

File: client/composeApp/src/commonMain/kotlin/net/treset/treelauncher/backend/launching/ResourceManager.kt
package net.treset.treelauncher.backend.launching

import io.github.oshai.kotlinlogging.KotlinLogging
import net.treset.treelauncher.backend.config.appConfig
import net.treset.treelauncher.backend.data.InstanceData
import net.treset.treelauncher.backend.data.manifest.ComponentManifest
import net.treset.treelauncher.backend.util.exception.GameResourceException
import net.treset.treelauncher.backend.util.file.LauncherFile
import net.treset.treelauncher.backend.util.string.PatternString
import java.io.IOException
import java.nio.file.StandardCopyOption
import java.time.LocalDateTime
import java.util.*

class ResourceManager(private var instanceData: InstanceData) {

    @Throws(GameResourceException::class)
    fun prepareResources() {
        try {
            instanceData.setActive(true)
        } catch (e: IOException) {
            throw GameResourceException("Failed to prepare resources: unable to set instance active", e)
        }
        addIncludedFiles(
            listOf(
                instanceData.instance.first,
                instanceData.optionsComponent,
                instanceData.resourcepacksComponent,
                instanceData.savesComponent
            )
        )
        instanceData.modsComponent?.let { addIncludedFiles(listOf(it.first)) }
        renameComponents()
        LOGGER.info {"Prepared resources for launch, instance=${instanceData.instance.first.id}"}
    }

    @Throws(IOException::class)
    fun setLastPlayedTime() {
        LOGGER.debug { "Setting last played time: instance=${instanceData.instance.first.id}" }
        val time = LocalDateTime.now()
        instanceData.instance.first.lastUsedTime = time
        instanceData.savesComponent.lastUsedTime = time
        instanceData.resourcepacksComponent.lastUsedTime = time
        instanceData.optionsComponent.lastUsedTime = time
        instanceData.modsComponent?.first?.lastUsedTime = time
        LauncherFile.of(
            instanceData.instance.first.directory,
            appConfig().manifestFileName
        ).write(instanceData.instance.first)
        LauncherFile.of(
            instanceData.savesComponent.directory,
            appConfig().manifestFileName
        ).write(instanceData.savesComponent)
        LauncherFile.of(
            instanceData.resourcepacksComponent.directory,
            appConfig().manifestFileName
        ).write(instanceData.resourcepacksComponent)
        LauncherFile.of(
            instanceData.optionsComponent.directory,
            appConfig().manifestFileName
        ).write(instanceData.optionsComponent)
        instanceData.modsComponent?.let {
            LauncherFile.of(
                it.first.directory,
                appConfig().manifestFileName
            ).write(it.first)
        }
        LOGGER.debug { "Set last played time: instance=${instanceData.instance.first.id}" }
    }

    @Throws(IOException::class)
    fun addPlayDuration(duration: Long) {
        LOGGER.debug { "Adding play duration: instance=${instanceData.instance.first.id}, duration=$duration" }

        val oldTime: Long = instanceData.instance.second.totalTime
        instanceData.instance.second.totalTime = oldTime + duration
        LauncherFile.of(
            instanceData.instance.first.directory,
            instanceData.instance.first.details
        ).write(instanceData.instance.second)
        LOGGER.debug { "Added play duration: instance=${instanceData.instance.first.id}, duration=$duration, totalTime=${instanceData.instance.second.totalTime}" }
    }

    @Throws(GameResourceException::class)
    fun cleanupGameFiles(mergeFiles: Boolean = false) {
        try {
            LOGGER.debug { "Cleaning up game files: instance=${instanceData.instance.first.id}" }
            try {
                undoRenameComponents()
                val gameDataFilesList = gameDataFiles
                removeExcludedFiles(gameDataFilesList)
                removeIncludedFiles(arrayOf(instanceData.savesComponent), gameDataFilesList, mergeFiles)
                instanceData.modsComponent?.let { removeIncludedFiles(arrayOf(it.first), gameDataFilesList, mergeFiles) }
                removeIncludedFiles(arrayOf(instanceData.optionsComponent, instanceData.resourcepacksComponent), gameDataFilesList, mergeFiles)
                removeIncludedFiles(instanceData.instance.first, gameDataFilesList, mergeFiles, true)
            } catch (e: GameResourceException) {
                throw GameResourceException("Unable to cleanup game files", e)
            }
            try {
                instanceData.setActive(false)
            } catch (e: IOException) {
                throw GameResourceException("Unable to cleanup game files: unable to set instance inactive", e)
            }
        } catch (e: Exception) {
            throw GameResourceException("Unable to cleanup game files", e)
        }
        LOGGER.info { "Game files cleaned up" }
    }

    @Throws(GameResourceException::class)
    private fun renameComponents() {
        LOGGER.debug { "Renaming components: instance=${instanceData.instance.first.id}" }
        try {
            val savesLocation = LauncherFile.of(instanceData.gameDataDir, "saves")
            LauncherFile.of(instanceData.savesComponent.directory).moveTo(savesLocation)
            instanceData.savesComponent.directory = savesLocation.path
        } catch (e: IOException) {
            throw GameResourceException("Unable to rename saves file", e)
        }
        instanceData.savesComponent.directory = LauncherFile.of(instanceData.gameDataDir, "saves").path
        instanceData.modsComponent?.let { modsComponents ->
            try {
                val modsLocation = LauncherFile.of(instanceData.gameDataDir, "mods")
                LauncherFile.of(modsComponents.first.directory).moveTo(modsLocation)
                modsComponents.first.directory = modsLocation.path
            } catch (e: IOException) {
                throw GameResourceException("Unable to rename mods file", e)
            }
            modsComponents.first.directory = LauncherFile.of(instanceData.gameDataDir, "mods").path
        }
    }

    @Throws(GameResourceException::class)
    private fun addIncludedFiles(components: List<ComponentManifest>) {
        LOGGER.debug { "Adding included files: instance=${instanceData.instance.first.id}" }
        val exceptionQueue: MutableList<GameResourceException> = mutableListOf()
        for (c in components) {
            try {
                addIncludedFiles(c)
            } catch (e: GameResourceException) {
                exceptionQueue.add(e)
                LOGGER.warn(e) { "Unable to get included files: manifestId=${c.id}"}
            }
        }
        if (exceptionQueue.isNotEmpty()) {
            throw GameResourceException("Unable to get included files for ${exceptionQueue.size} components", exceptionQueue[0])
        }
        LOGGER.debug { "Added included files: instance=${instanceData.instance.first.id}" }
    }

    @Throws(GameResourceException::class)
    private fun addIncludedFiles(manifest: ComponentManifest) {
        val includedFilesDir: LauncherFile = LauncherFile.of(manifest.directory, appConfig().includedFilesDirName)
        if (!includedFilesDir.isDirectory()) {
            try {
                includedFilesDir.createDir()
            } catch (e: IOException) {
                throw GameResourceException("Unable to get included files: unable to create included files directory: manifestId=${manifest.id}", e)
            }
        }
        val files = includedFilesDir.listFiles()
        LOGGER.debug { "Adding included files: ${manifest.type}, id=${manifest.id}, includedFiles=$files" }
        val exceptionQueue: MutableList<Exception> = ArrayList<Exception>()
        for (f in files) {
            LOGGER.debug { "Moving file: ${f.name}" }
            if (f.isFile() || f.isDirectory()) {
                try {
                    LauncherFile.of(f).copyTo(
                        LauncherFile.of(instanceData.gameDataDir, f.getName()),
                        StandardCopyOption.REPLACE_EXISTING
                    )
                } catch (e: Exception) {
                    exceptionQueue.add(e)
                    LOGGER.warn(e) { "Unable to move included files: unable to copy file: manifestId=${manifest.id}" }
                }
            } else {
                exceptionQueue.add(IOException("Included files directory contains invalid file type: manifestId=${manifest.id}"))
            }
        }
        if (exceptionQueue.isNotEmpty()) {
            throw GameResourceException("Unable to move included files: unable to copy ${exceptionQueue.size} files", exceptionQueue[0])
        }
        LOGGER.debug { "Added included files: manifestId=${manifest.id}}" }
    }

    @Throws(GameResourceException::class)
    private fun undoRenameComponents() {
        LOGGER.debug { "Undoing component renames: instance=${instanceData.instance.first.id}" }
        val newSavesDir = LauncherFile.ofData(instanceData.launcherDetails.savesDir, "${instanceData.savesPrefix}_${instanceData.savesComponent.id}")
        try {
            LauncherFile.of(instanceData.savesComponent.directory).moveTo(newSavesDir)
            instanceData.savesComponent.directory = newSavesDir.path
        } catch (e: IOException) {
            throw GameResourceException("Unable to cleanup launch resources: rename saves file failed", e)
        }
        instanceData.savesComponent.directory = newSavesDir.path
        instanceData.modsComponent?.let { modsComponents ->
            val newModsDir = LauncherFile.ofData(instanceData.launcherDetails.modsDir, "${instanceData.modsPrefix}_${modsComponents.first.id}")
            try {
                LauncherFile.of(modsComponents.first.directory).moveTo(newModsDir)
                instanceData.modsComponent?.first?.directory = newModsDir.path
            } catch (e: IOException) {
                throw GameResourceException("Unable to cleanup launch resources: rename mods file failed", e)
            }
            modsComponents.first.directory = newModsDir.path
        } ?: run {
            LOGGER.debug { "No mods component to cleanup, deleting mods dir" }
            val modsDir = LauncherFile.of(instanceData.gameDataDir, "mods")
            if (modsDir.exists()) {
                try {
                    modsDir.remove()
                } catch (e: IOException) {
                    throw GameResourceException("Unable to cleanup launch resources: unable to delete mods directory", e)
                }
            }
        }
        LOGGER.debug { "Undid component renames: instance=${instanceData.instance.first.id}" }
    }

    @get:Throws(GameResourceException::class)
    private val gameDataFiles: MutableList<LauncherFile>
        get() {
            LOGGER.debug { "Getting game data files: instance=${instanceData.instance.first.id}" }
            val gameDataDir: LauncherFile = instanceData.gameDataDir
            if (!gameDataDir.isDirectory()) {
                throw GameResourceException("Unable to cleanup launch resources: game data directory not found")
            }
            val gameDataFiles = gameDataDir.listFiles()
            LOGGER.debug { "Got game data files: instance=${instanceData.instance.first.id}" }
            return gameDataFiles.toMutableList()
        }

    private fun removeExcludedFiles(files: MutableList<LauncherFile>) {
        LOGGER.debug { "Removing excluded files: instance=${instanceData.instance.first.id}, files=$files" }
        val toRemove: MutableList<LauncherFile> = mutableListOf()
        for (f in files) {
            val launcherName = f.getLauncherName()

            if (launcherName == appConfig().manifestFileName || PatternString.matchesAny(launcherName, instanceData.gameDataExcludedFiles)) {
                LOGGER.debug { "Removing excluded file: ${f.name}" }
                toRemove.add(f)
            }
        }
        files.removeAll(toRemove.toSet())
        LOGGER.debug { "Removed excluded files: instance=${instanceData.instance.first.id}" }
    }

    @Throws(GameResourceException::class)
    private fun removeIncludedFiles(components: Array<ComponentManifest>, files: MutableList<LauncherFile>, mergeFiles: Boolean) {
        LOGGER.debug { "Removing included files: instance=${instanceData.instance.first.id}, files=$files" }
        val exceptionQueue: MutableList<GameResourceException> = mutableListOf()
        for (component in components) {
            if (component.includedFiles.isEmpty()) {
                LOGGER.debug { "No included files: ${component.type.name.lowercase()}, manifestId=${component.id}" }
                continue
            }
            try {
                removeIncludedFiles(component, files, mergeFiles)
            } catch (e: GameResourceException) {
                exceptionQueue.add(e)
                LOGGER.warn(e) { "Unable to remove included files: ${component.type.name.lowercase()}, manifestId=${component.id}" }
            }
        }
        if (exceptionQueue.isNotEmpty()) {
            throw GameResourceException("Unable to remove included files for ${exceptionQueue.size} components", exceptionQueue[0])
        }
        LOGGER.debug { "Removed included files: instance=${instanceData.instance.first.id}}" }
    }

    @Throws(GameResourceException::class)
    private fun removeIncludedFiles(component: ComponentManifest, files: MutableList<LauncherFile>, mergeFiles: Boolean, allFiles: Boolean = false) {
        LOGGER.debug { "Removing included files: ${component.type.name.lowercase()}, id=${component.id}, includedFiles=${component.includedFiles}, files=$files" }
        val includedFilesDir: LauncherFile =
            LauncherFile.of(component.directory, appConfig().includedFilesDirName)
        val oldIncludedFilesDir: LauncherFile =
            LauncherFile.of(component.directory, "${appConfig().includedFilesDirName}_old")
        if (includedFilesDir.exists()) {
            try {
                if (oldIncludedFilesDir.exists()) {
                    oldIncludedFilesDir.remove()
                }
                if(mergeFiles) {
                    includedFilesDir.copyTo(oldIncludedFilesDir, StandardCopyOption.REPLACE_EXISTING)
                } else {
                    includedFilesDir.moveTo(oldIncludedFilesDir, StandardCopyOption.REPLACE_EXISTING)
                }
            } catch (e: IOException) {
                throw GameResourceException("Unable to remove included files: unable to move included files directory: component_type=${component.type.name.lowercase()}, component=${component.id}", e)
            }
        }
        try {
            includedFilesDir.createDir()
        } catch (e: IOException) {
            throw GameResourceException("Unable to remove included files: unable to create included files directory: component_type=${component.type.name.lowercase()}, component=${component.id}" )
        }

        val currentFiles = component.includedFiles.map { p -> PatternString(p, true).changeDirectoryEnding() }.toTypedArray()

        val toRemove: MutableList<LauncherFile> = mutableListOf()
        val exceptionQueue: MutableList<IOException> = mutableListOf()
        for (f in files) {
            if(allFiles || PatternString.matchesAny(f.getLauncherName(), currentFiles)) {
                toRemove.add(f)

                LOGGER.debug { "Moving file: ${f.name}" }
                try {
                    f.moveTo(LauncherFile.of(component.directory, appConfig().includedFilesDirName, f.getName()), StandardCopyOption.REPLACE_EXISTING)
                } catch (e: IOException) {
                    exceptionQueue.add(e)
                    LOGGER.warn { "Unable to remove included files: unable to move file: component_type=${component.type.name.lowercase()}, component=${component.id}, file=${f}" }
                }
            }
        }
        files.removeAll(toRemove.toSet())
        if (exceptionQueue.isNotEmpty()) {
            throw GameResourceException("Unable to remove included files: unable to move ${exceptionQueue.size} files: component_type=${component.type.name.lowercase()}, component=${component.id}", exceptionQueue[0])
        }
        LOGGER.debug { "Removed included files: manifestId=${component.id}}" }
    }

    companion object {
        private val LOGGER = KotlinLogging.logger {}
    }
}
